### ISSUE TITLE
Misc. explore mode slowdowns

### ISSUE THREAD

**Issue by run2cmd** [Jelmerro/Vieb#387]
**Checklist**
- [x] I searched the `:help` documentation using `/` for helpful information
- [x] I have read the [frequently asked questions](https://github.com/Jelmerro/Vieb/blob/master/FAQ.md)
- [x] I did not find any similar issues in the [issue tracker](https://github.com/Jelmerro/Vieb/issues)

**Describe the bug**
I'll describe few issues here I've encountered with `Explore` mode. I did not open Issue for each one as they might be related and all of them regard `Explore` mode.

1. I have noticed slowdown with `Explore` mode. At Vieb startup when I open new tab, press `e` for explore mode and start writing Vieb hangs for 15-20 seconds, before inserting text. During this period I can wring what I need but you don't see any text typed and hitting `enter` button will go to URL after that period is complete. This happens only with `Explore` mode as I can use `Command` mode just after Vieb starts.

2. Another issue  is also with `Explore` not going to website after I start typing URL and choose one of the URL on the popup list to go to. It sometimes does not go to that URL and I need to press `e<CR>` few times to get there, and in some occasions it even makes tab unusable (tab color goes red) and I need to open new tab.

3. Last one I've noticed is that when I type in `Explore` mode and use `backspace` button to delete text it causes Vieb to work slow. Sometimes I need to wait few seconds before text is remove so I can alter it properly. This is mostly happen when you hold `backspace` to delete.

I'm happy to help with debugging in case there is anything I can do on PC to help narrow down the issue. I just do not know how to do that.

**To Reproduce**

1. Open new Vieb instance, press `te` and start typing.
2. This might be hard to reproduce as this seems to be happening to me randomly. To reproduce basically you do `te<write-text><tab><cr>`
3. `te<write-text><tab><hold-backspace>`

**Expected behavior**

For all issues it should not slowdown and actions should be executed immediately. 

**Specs**
 - OS: Windows
 - Version: 10

**Comment by Jelmerro** [comment]
These are actually 2 issues, only 1 of which can be solved, allow me to explain them 1 by 1.

The first issue is not Windows specific, but it's just a lot worse on Windows, because file access is slow beyond believe on that system. Opening a new tab takes milliseconds on most systems, but on Windows Vieb has to wait to read files before this can happen. If you then immediately start typing, it's not a big issue on other platforms, but it's notable on Windows because the interface freezes for a moment. This freeze period is then extended if you start typing right away, hence giving this issue. But as said, this is only one part of it, as typing in itself should not be enough to freeze the window, even for a moment.

This brings us to the 2nd issue, which has to do with the way explore mode suggestions are generated in Vieb. As you can see [in the git blame of the history file](https://github.com/Jelmerro/Vieb/blame/master/app/renderer/history.js#L44) this is fairly old code, but I have personally in all this time not found many ways to improve it's speed while keeping it's functionality intact. For this task, I have also looked at different string matching libraries in the npm ecosystem. I must have tested around 10 libraries, NONE of which were faster than the current implementation, while ALL of them had worse results. Most libraries also had no way of integrating the visit count on top of the relevance, which leads to very biased results either toward visit count or relevance, which a mix of them is desired for explore suggestions. I have recently been looking into the performance of this function again, and noticed that most of the time is spend on simplifying the titles, as was the case for urls before we introduced the `simpleUrls` cache object. As such, I might be able to improve the speed by caching the page titles. Another method I recently thought of, which would also be nice to do on the history page, is to limited the history to a specific period, such as past year, as to reduce the total amount of history that has to be parsed. 

Overall, this entire issue is a combination of having a lot of history and file access being slow in Windows, neither of which can be easily solved, but I will work on a fix for the page title caching. Please let me know if you have additional ideas or comments about this, especially regarding the long-term status of this issue and what else we could do to solve it.
![image](https://user-images.githubusercontent.com/1696674/176416211-aa279c65-b4c0-4bb6-8817-9b9adbdf938d.png)

**Comment by Jelmerro** [comment]
I managed to get [page title caching working](https://github.com/Jelmerro/Vieb/commit/fc5d9666e2ecf0bbbde15ec225c042a80cc5d229), so while the first search could be a bit slower, any subsequent search will use this cache and thus be faster. Let me know if this is better for you by running Vieb from source (as per readme), and let me know what else you would like to see changed to resolve this issue.

**Comment by Jelmerro** [comment]
After some performance tests I decided to make a new release, which includes this fix, feel free to test it with that. It's a bit slower the first time, but much faster on subsequent searches.

**Comment by run2cmd** [comment]
@Jelmerro thanks for deep explanation of the problem :). I've run some tests on latest version and is seems to be better. As issue #1 still remains problem #2 is much better now. The 1st time start is something I can live with since I tend to use instance for hours. Unfortunately #3 is still there especially when you hold `backspace` button.

If the issue is really about Windows file access (I understand this is Hard Drive file access) then the only option is to move it to RAM. I'm not sure if you considered loading history to memory(as some variable maybe) and dump to file on browser quit so history is preserved. I know that there is downside of higher memory usage but if there would be option to limit history to certain (set in configuration) number of pages then this could be better controlled. This all would be optional for people (like me) that have to work on Windows.

**Comment by Jelmerro** [comment]
Just to be clear, the history is already stored in memory once it's loaded, the only thing that is significantly slower on windows is opening new tabs. If it's still not fast enough on your computer the only solution I can think of is an option to limit explore mode to the recent history (and "recent" being configurable with a setting).

**Comment by nitanmarcel** [comment]
@Jelmerro should the 2nd issue be fixed in `8.1.0` ? Because I get the same thing with it. Sometimes I have to press e and enter multiple time to go to a different URL.

**Comment by Jelmerro** [comment]
Which issue are you referring to? In the latest release both urls and title simplifications are cached after generating them once, so the first search might be slower, but later ones should be faster now. If it's still slow, the only thing that's going to help is limited the amount of entries that are search with an option to limit it to recent history, for which there isn't a setting as of yet, hence this issue still being open.

**Comment by nitanmarcel** [comment]
> 2\. Another issue  is also with `Explore` not going to website after I start typing URL and choose one of the URL on the popup list to go to. It sometimes does not go to that URL and I need to press `e<CR>` few times to get there, and in some occasions it even makes tab unusable (tab color goes red) and I need to open new tab.

The one mentioned at the point 2, and quoted bellow.

> Another issue is also with Explore not going to website after I start typing URL and choose one of the URL on the popup list to go to. It sometimes does not go to that URL and I need to press e<CR> few times to get there, and in some occasions it even makes tab unusable (tab color goes red) and I need to open new tab.

![image](https://user-images.githubusercontent.com/41646249/179461827-e419da05-42cc-428a-b0d7-0f4961f91257.png)

**Comment by run2cmd** [comment]
I build Vieb from latest commit 3dc00bdb6946906d4e94e5d42b221f50246e00f9 and I can confirm that issue 1 and 3 are solved. There is still slowdown as @Jelmerro mentioned with 1st time entering explore mode, but it's much shorter. Basically explore mode works blazingly fast comparing to 8.1.0 :).

Unfortunately there is still issue 2 with certain websites loading. I mostly get that on websites like youtube, twitter, bank sites etc. Ones that I do not enter very often during day. After entering site I still need to do `:e<cr>` at least twice (sometimes 4-5 times) to get to site. Here is error I got (It is the same for other sites, just url changes):

```
[28072:0720/080232.087:ERROR:node_bindings.cc(504)] Failed to initialize node environment in process: worker , with error: Uncaught Error: No such module: native_module
Error occurred in handler for 'GUEST_VIEW_MANAGER_CALL': Error: ERR_FAILED (-2) loading 'https://www.youtube.com/'
    at rejectAndCleanup (node:electron/js2c/browser_init:161:7647)
    at EventEmitter.stopLoadingListener (node:electron/js2c/browser_init:161:8022)
    at EventEmitter.emit (node:events:539:35) {
  errno: -2,
  code: 'ERR_FAILED',
  url: 'https://www.youtube.com/'
}
```

**Comment by nitanmarcel** [comment]
> I build Vieb from latest commit [3dc00bd](https://github.com/Jelmerro/Vieb/commit/3dc00bdb6946906d4e94e5d42b221f50246e00f9) and I can confirm that issue 1 and 3 are solved. There is still slowdown as @Jelmerro mentioned with 1st time entering explore mode, but it's much shorter. Basically explore mode works blazingly fast comparing to 8.1.0 :).
> 
> Unfortunately there is still issue 2 with certain websites loading. I mostly get that on websites like youtube, twitter, bank sites etc. Ones that I do not enter very often during day. After entering site I still need to do `:e<cr>` at least twice (sometimes 4-5 times) to get to site. Here is error I got (It is the same for other sites, just url changes):
> 
> ```
> [28072:0720/080232.087:ERROR:node_bindings.cc(504)] Failed to initialize node environment in process: worker , with error: Uncaught Error: No such module: native_module
> Error occurred in handler for 'GUEST_VIEW_MANAGER_CALL': Error: ERR_FAILED (-2) loading 'https://www.youtube.com/'
>     at rejectAndCleanup (node:electron/js2c/browser_init:161:7647)
>     at EventEmitter.stopLoadingListener (node:electron/js2c/browser_init:161:8022)
>     at EventEmitter.emit (node:events:539:35) {
>   errno: -2,
>   code: 'ERR_FAILED',
>   url: 'https://www.youtube.com/'
> }
> ```

The second issue happens anywhere I suppose. I was browsing reddit the other day and I had to spam the explorer mode to go to the website. This also happens when manually trying to set an url to go to.

**Comment by run2cmd** [comment]
I did some testing and noticed few things about 2nd issue.

1. The issue persisted no matter what cache settings I setup.
2. The problem happens only on 1st time when I open page. So when I open https://youtube.com once I do not experience 2nd issue to happen within same session.
3. As last resort I have removed my data directory (`c:\Users\<user>\AppData\Roaming\Vieb`) and I do not experience any issues with entering sites.

Now the 3rd step was done quite recently but on 2-3 days I should fill data directory with enough data to confirm that this solve the issue. Would be good to know what actually would be best to clean (history only maybe?) to narrow down the clean up as this problem might be related to locally stored data corruption or some change/update that affected performance with old data.

During this testing I also found interesting thing about 1st issue. When I `set restoretabs=false` it actually got rid of the problem on last released version (8.1.0).

**Comment by nitanmarcel** [comment]
> Myself even if I `set restoretabs=false` I get the issue

**Comment by nitanmarcel** [comment]
@Jelmerro as @run2cmd said above a few comments this is what the logs say when a page doesn't load the first time.

I tried with multiple websites I know I can reproduce the issue.

```
[10044:0725/141951.075:ERROR:node_bindings.cc(504)] Failed to initialize node environment in process: worker , with error: Uncaught Error: No such module: native_module
Error occurred in handler for 'GUEST_VIEW_MANAGER_CALL': Error: ERR_FAILED (-2) loading 'https://www.youtube.com/'
    at rejectAndCleanup (node:electron/js2c/browser_init:161:7647)
    at EventEmitter.stopLoadingListener (node:electron/js2c/browser_init:161:8022)
    at EventEmitter.emit (node:events:539:35) {
  errno: -2,
  code: 'ERR_FAILED',
  url: 'https://www.youtube.com/'
}
[25168:0725/142019.000:ERROR:node_bindings.cc(504)] Failed to initialize node environment in process: worker , with error: Uncaught Error: No such module: native_module
Error occurred in handler for 'GUEST_VIEW_MANAGER_CALL': Error: ERR_ABORTED (-3) loading 'https://www.google.com/url?esrc=s&q=&rct=j&sa=U&url=https://www.reddit.com/r/funny/&ved=2ahUKEwiYw5by9ZP5AhVynP0HHaUtDPkQFnoECAgQAg&usg=AOvVaw0Pn4UmDlMIQ3XBeR92cFMA'
    at rejectAndCleanup (node:electron/js2c/browser_init:161:7647)
    at EventEmitter.navigationListener (node:electron/js2c/browser_init:161:7953)
    at EventEmitter.emit (node:events:539:35) {
  errno: -3,
  code: 'ERR_ABORTED',
  url: 'https://www.google.com/url?esrc=s&q=&rct=j&sa=U&url=https://www.reddit.com/r/funny/&ved=2ahUKEwiYw5by9ZP5AhVynP0HHaUtDPkQFnoECAgQAg&usg=AOvVaw0Pn4UmDlMIQ3XBeR92cFMA'
}
[13232:0725/142026.187:ERROR:node_bindings.cc(504)] Failed to initialize node environment in process: worker , with error: Uncaught Error: No such module: native_module
```

**Comment by run2cmd** [comment]
After few days of testing unfortunately problem 2 returned. I tried to install `native-module` but it failed on make so I did not dig up more to that. Also `native-module` seems to required `node-gyp` which has a lot of dependencies (python, c++). At this point I'm not sure what else I can do.

**Comment by Jelmerro** [comment]
Most of these issues have been fixed in 9.0.0, except for the first tab load issue, where I have no clue as of yet what is going wrong.

**Comment by Jelmerro** [comment]
I have made a slight improvement to fix it for newly opened tabs, but for existing tabs that you use explore mode for this can still happen. I also found the matching Electron issue, as this is not something that is Vieb specific: https://github.com/electron/electron/issues/28208

**Comment by Jelmerro** [comment]
This was a tough one, but this specific issue appears to be fixed by Electron and I have released 9.4.0 containing the new version. If you have any tab loading or explore more issues in the future, just let me know, or preferably open a new issue for it with updated instructions.
